### PR TITLE
Add attention kernel integral layer

### PR DESCRIPTION
**Pull request by zijieli-Jlee**
This pull request adds attention kernel integral layer, which uses attention to compute the learnable kernel integral.

File added is: ```layers/attention_kernel_integral.py```

### CODE REVIEW

File: neuralop/layers/attention_kernel_integral.py [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+        # initialization for the projection matrix
+        # basically initialize the weights for each heads as
+        # this requires input dim = dim_head
+        # W = init_fn(W) + I * diagonal_weight
+        # init_fn is usually xavier_uniform_
+        for param in weight.parameters():
+            if param.ndim > 1:
+                dim_head = param.size(0) // self.heads
+                for h in range(self.heads):
+                    inif_fn(param[h * dim_head:(h + 1) * dim_head, :], gain=self.init_gain)
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:70]: Something went wrong with formatting here

File: neuralop/layers/attention_kernel_integral.py [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+    use_pe : bool, whether to use positional encoding
+    project_query : bool, whether to project the query function with pointwise linear layer
+    """
+
+    def __init__(self,
+                 channels,
+                 heads,
+                 dim_head,
+                 pos_dim,
+                 use_pe=True,    # use positional encoding
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:30]: Can we make the variable names mode explicit? 
e.g. `n_heads`, `head_n_channels` (or dim is fine too), `use_positional_encoding` etc?

Comment (JeanKossaifi) [reply neuralop/layers/attention_kernel_integral.py:30]: Will be easier for users not familiar with the architecture

File: neuralop/layers/attention_kernel_integral.py [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        self.init_weight(self.to_v, init_fn)
+
+    def normalize_wrt_domain(self, x, norm_fn):
+        b = x.shape[0]
+        x = x.view(b*self.heads, -1, self.dim_head)
+        x = norm_fn(x)
+        return x.view(b, self.heads, -1, self.dim_head)
+
+    def forward(self,
+                u_x,
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:92]: Same here, could we find a good name, e.g just `x` and `y` for consistency with rest or something else? Though we could also call all inputs `u_x` in the code or just `f_x`

File: neuralop/layers/attention_kernel_integral.py [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+                pos_y=None,
+                weights=None,
+                associate=True,   # can be much faster if n is larger than the channel number c
+                get_kernel=False):
+        """
+        Computes kernel integral transform with attention
+
+        Parameters
+        ----------
+        u_x: input (query) function of shape [b, n, c]
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:105]: Can we expand b, n c to full variable name?

File: neuralop/layers/attention_kernel_integral.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        pos_x: coordinate of input function's grid points [b, n, d]
+        pos_emb: positional encoding module for encoding q/k torch.nn.Module
+        u_y: the second source of function (key and value), if not provided, u_y = u_x
+        pos_y: coordinate of the second source of function's grid points, if not provided, assume pos_y = pos_x
+        weights : tensor of shape [b, n], if not provided assume to be 1/n
+                  Weights for each point y proprtional to the
+                  volume around f(y)=u_y W_v being integrated.
+        associate: if True, use associativity of matrix multiplication, first multiply K^T V, then multiply Q
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:113]: How about `associative`?

**Review by JeanKossaifi — COMMENTED** [review]
Thanks @zijieli-Jlee , it looks good! I had a quick look and left some comments.

File: neuralop/layers/attention_kernel_integral.py (hunk 1 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Reformatted this part, let me know if the issues remain

File: neuralop/layers/attention_kernel_integral.py (hunk 2 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Sure, change the variable name to make them more explicit

File: neuralop/layers/attention_kernel_integral.py (hunk 3 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py:91]: With the new commit, I renamed `x` to `u`, I feel `u` can be more general as `u_x` and `u_y` both will probably be applied normalization (though `u_x` is not being normalized in the current code)

File: neuralop/layers/attention_kernel_integral.py (hunk 4 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Done

File: neuralop/layers/attention_kernel_integral.py (hunk 5 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Changed

File: neuralop/layers/attention_kernel_integral.py [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+    """
+    Kernel integral transform with attention
+    Computes \int_{Omega} k(x, y) * f(y) dy,
+    where:
+          K(x, y) = \sum_{c=1}^d \q_c(x) * \k_c(y)
+          f(y) = v(y)
+
+    Parameters
+    ----------
+    channels : int, input and output channels
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:17]: Do we not want to have allow them to be different? 
Nitpick: for consistency with other layers would be great to call then either in_channels and out_channels or just `n_channels`.

File: neuralop/layers/attention_kernel_integral.py [hunk 7]
```diff
@@ -0,0 +1,194 @@
...
+    n_heads : int, number of attention n_heads
+    head_n_channels : int, dimension of each attention head, determines how many function bases to use for the kernel
+    pos_dim : int, dimension of domain
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:20]: Not clear to me what this is from this description - could we name it more explicitly or expand?

File: neuralop/layers/attention_kernel_integral.py [hunk 8]
```diff
@@ -0,0 +1,194 @@
...
+
+    def initialize_qkv_weights(self):
+        init_fn = xavier_uniform_
+
+        if self.project_query:
+            self.init_weight(self.to_q, init_fn)
+        self.init_weight(self.to_k, init_fn)
+        self.init_weight(self.to_v, init_fn)
+
+    def normalize_wrt_domain(self, u, norm_fn):
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:83]: A small docstring would be great :)

File: neuralop/layers/attention_kernel_integral.py (hunk 3 above)
Comment (JeanKossaifi) [reply neuralop/layers/attention_kernel_integral.py:91]: I like the new notation with u_x and u_y!

File: neuralop/layers/attention_kernel_integral.py [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+        # u: the input or transformed function
+        batch_size = u.shape[0]
+        u = u.view(batch_size*self.n_heads, -1, self.head_n_channels)
+        u = norm_fn(u)    # layer norm with channel dimension or instance norm with spatial dimension
+        return u.view(batch_size, self.n_heads, -1, self.head_n_channels)
+
+    def forward(self,
+                u_x,
+                pos_x,
+                pos_emb=None,    # positional encoding module for encoding q/k
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:93]: Can we spell out the variables, e.g. here pos_embedding or even positional_embedding? It's a little longer to type but makes code very readable

File: neuralop/layers/attention_kernel_integral.py [hunk 10]
```diff
@@ -0,0 +1,194 @@
...
+                u_y=None,
+                pos_y=None,
+                weights=None,
+                associative=True,   # can be much faster if num_grid_points is larger than the channel number c
+                get_kernel=False):
```
Comment (JeanKossaifi) [inline neuralop/layers/attention_kernel_integral.py:98]: nitpick: `return_kernel` is clearer

**Review by JeanKossaifi — COMMENTED** [review]
Thanks a lot @zijieli-Jlee, this is looking good! @crwhite14 for a quick review if you have time.

File: neuralop/layers/attention_kernel_integral.py (hunk 6 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Yes definitely and that's more reasonable. I often omit it since input and output channels in attention are almost always the same to accommodate the residual

File: neuralop/layers/attention_kernel_integral.py (hunk 7 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: I expanded it a bit more

File: neuralop/layers/attention_kernel_integral.py (hunk 8 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py:96]: Added a small doctring here

File: neuralop/layers/attention_kernel_integral.py (hunk 9 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: I named it positional_embedding_module, if it's too lengthy I can also change it to positional_embedding or sth else

File: neuralop/layers/attention_kernel_integral.py (hunk 10 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Certainly

File: neuralop/layers/attention_kernel_integral.py [hunk 11]
```diff
@@ -0,0 +1,212 @@
...
+        self.k_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+        self.v_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+
+        self.to_v = nn.Linear(in_channels, head_n_channels * n_heads, bias=False)
+
+        self.to_out = nn.Linear(head_n_channels * n_heads, out_channels) \
+            if head_n_channels * n_heads != out_channels else nn.Identity()
+
+        self.use_positional_encoding = use_positional_encoding
+        self.pos_dim = pos_dim
```
Comment (juliusberner) [inline neuralop/layers/attention_kernel_integral.py:59]: Why do we need `self.pos_dim`? We can just infer it from `pos_x` at runtime.

File: neuralop/layers/attention_kernel_integral.py [hunk 12]
```diff
@@ -0,0 +1,212 @@
...
+        Parameters
+        ----------
+        u_x: input (query) function of shape [batch_size, num_grid_points, channels]
+        pos_x: coordinate of input function's grid points [batch_size, num_grid_points, pos_dim]
+        positional_embedding_module: positional embedding module for encoding query/key (q/k), a torch.nn.Module
+        u_y: the second source of function (key and value), if not provided, u_y = u_x
+        pos_y: coordinate of the second source of function's grid points, if not provided, assume pos_y = pos_x
+        weights : tensor of shape [batch_size, num_grid_points], if not provided assume to be 1/num_grid_points
+                  Weights for each point y proportional to the
+                  volume around f(y)=u_y W_v being integrated.
```
Comment (juliusberner) [inline neuralop/layers/attention_kernel_integral.py:129]: It is unclear what `W_v` here is.

**Review by juliusberner — CHANGES_REQUESTED** [review]

File: neuralop/layers/attention_kernel_integral.py [hunk 13]
```diff
@@ -0,0 +1,212 @@
...
+from torch.nn.init import xavier_uniform_, zeros_
+
+
+class AttentionKernelIntegral(torch.nn.Module):
+    """
+    Kernel integral transform with attention
+    Computes \int_{Omega} k(x, y) * f(y) dy,
+    where:
+          K(x, y) = \sum_{c=1}^d \q_c(x) * \k_c(y)
+          f(y) = v(y)
```
[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py:13]: Thanks for the comment here. My previous docstring and variable naming might be a bit vague. I added a more detailed description and renamed some variables in the `forward` in the latest commit.

Here is what I added to the docstring and I also renamed variable based on this: 
This module computes $\int_{\Omega} k(x, y) * f(y) dy$,
with:
      $$f(y)=v(y), k(x, y) = \sum_{c=1}^d q_c(x) * k_c(y), \text{where: } q(x) = [q_1(x); ...; q_d(x)], k(y) = [k_1(y); ...;k_d(y)]$$
The learnable weights are $W_q, W_k, W_v \in \mathbb{R}^{d \times d}$.
1. Self-attention:
    There is only a single input function $u$ and its grid $D_x$
    query function: $q(x_i) = a(x_i) W_q$
    key function: $k(x_i) = a(x_i) W_k$
    value function: $v(x_i) = a(x_i) W_v$

2. Cross-attention:
    First input function $u_{qry}$, and its grid $D_x$
    Second input function $u_{src}$, and its grid $D_y$.
    query function: $q(x_i) = u_{qry}(x_i) W_q$
    key function: $k(y_j) = u_{src}(y_j) W_k$
    value function: $v(y_j) = u_{src}(y_j) W_v$

Self-attention can be considered as a special case of cross-attention, where $u = u_{qry} = u_{src}$ and $D_x = D_y$.

File: neuralop/layers/attention_kernel_integral.py [hunk 14]
```diff
@@ -0,0 +1,212 @@
...
+
+        self.k_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+        self.v_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+
+        self.to_v = nn.Linear(in_channels, head_n_channels * n_heads, bias=False)
+
+        self.to_out = nn.Linear(head_n_channels * n_heads, out_channels) \
+            if head_n_channels * n_heads != out_channels else nn.Identity()
+
+        self.use_positional_encoding = use_positional_encoding
```
[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Yes, removed

File: neuralop/layers/attention_kernel_integral.py [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+                associative=True,   # can be much faster if num_grid_points is larger than the channel number c
+                return_kernel=False):
+        """
+        Computes kernel integral transform with attention
+
+        Parameters
+        ----------
+        u_x: input (query) function of shape [batch_size, num_grid_points, channels]
+        pos_x: coordinate of input function's grid points [batch_size, num_grid_points, pos_dim]
+        positional_embedding_module: positional embedding module for encoding query/key (q/k), a torch.nn.Module
```
[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Currently I haven't pushed that part to the library.
In a preliminary demo I showed Jean I import the positional embedding module first and construct it in `__init__` (that's why there's some dummy variable related to positional encoding defined left over in the init of the class). Jean suggests passing the embedding class directly to the layer could be potentially more flexible. I can change it such that the Embedding class is passed in during initializing the attention layer and becomes an attribute.

File: neuralop/layers/attention_kernel_integral.py [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+                 in_channels,
+                 out_channels,
+                 n_heads,
+                 head_n_channels,
+                 pos_dim,
+                 use_positional_encoding=True,    # use positional encoding
+                 project_query=True,
+                 ):
+        super().__init__()
+        self.layers = nn.ModuleList([])
```
[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Removed

File: neuralop/layers/attention_kernel_integral.py [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+        q = q.view(batch_size, -1, self.n_heads, self.head_n_channels).permute(0, 2, 1, 3).contiguous()
+        k = k.view(batch_size, -1, self.n_heads, self.head_n_channels).permute(0, 2, 1, 3).contiguous()
+        v = v.view(batch_size, -1, self.n_heads, self.head_n_channels).permute(0, 2, 1, 3).contiguous()
+
+        if weights is None:
+            weights = torch.ones((u_y.shape[0], 1, u_y.shape[1], 1), device=u_y.device) / num_grid_points   # uniform weights
+        else:
+            weights = weights.view(batch_size, 1, num_grid_points, 1)
+
+        # q = self.q_norm(q)
```
[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Yes

File: neuralop/layers/attention_kernel_integral.py (hunk 11 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Yes

File: neuralop/layers/attention_kernel_integral.py (hunk 12 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Elaborated a bit more in the new docstring, it's a linear layer projecting the input to value in attention

File: neuralop/layers/attention_kernel_integral.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (juliusberner) [reply neuralop/layers/attention_kernel_integral.py]: I think both options are fine! But it would be great to add the embedding to this pull request since this layer is not usable otherwise.

File: neuralop/layers/attention_kernel_integral.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (juliusberner) [reply neuralop/layers/attention_kernel_integral.py:13]: Great, thanks a lot for the quick update! Two nits:
1. I think the `a` should be a `u` in the first item.
2. We could add the discretization `sum_{i} k(x_i, x_i) f(x_i) w(x_i)` (and analogously for the cross-attention)  where we mention the quadrature weights `w(x_i)`.
2. For self-attention, the `forward` currently seems to support `D_y` to be different from `D_x`.  I think this should raise a `ValueError` since we do not have the matching function values in this case.

File: neuralop/layers/attention_kernel_integral.py [hunk 18]
```diff
@@ -0,0 +1,228 @@
...
+
+        Output
+        ----------
+        u: Output function given on the query points x.
+        """
+
+        if u_qry is None:
+            u_qry = u_src   # go back to self attention
+        else:
+            assert pos_qry is not None, 'query coordinates are required if query function is provided'
```
Comment (juliusberner) [inline neuralop/layers/attention_kernel_integral.py:158]: Can we raise a `ValueError` here and also for the `Exception` below?

File: neuralop/layers/attention_kernel_integral.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Sure, I will add an embedding layer to this pull request

File: neuralop/layers/attention_kernel_integral.py (hunk 18 above)
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Yes

File: neuralop/layers/attention_kernel_integral.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (juliusberner) [reply neuralop/layers/attention_kernel_integral.py]: Great, thanks in advance :)

[the remark this answers is not in the document]
Comment (zijieli-Jlee) [reply neuralop/layers/attention_kernel_integral.py]: Added rotary positional encoding under `embeddings.py`

**Review by juliusberner — APPROVED** [review]
I left two minor comments on RoPE. Otherwise LGTM. 
Perhaps, we can test the forward of the attention layer with RoPE.
